src/core/ddsi/src/q_plist.c

```diff
@@ -2768,44 +2768,62 @@ static void fill_property(dds_property_t *prop, const char *name, const char *va
   prop->propagate = false;
 }
 
-void nn_xqos_mergein_security_config (dds_qos_t *xqos, const struct omg_security_configuration_type *cfg)
+bool nn_xqos_mergein_security_config (dds_qos_t *xqos, const struct omg_security_configuration_type *cfg)
 {
-  if (cfg)
-  {
-    /* Add DDS Security configuration to the QoS as a Property policy,
-     * used by the security plugins to get their proper settings. */
-    dds_property_t *properties;
-    uint32_t idx = 0;
+  bool qos_cfg = false;
+  assert(cfg != NULL);
 
-    /* No property should be present yet. */
-    assert(!(xqos->present & QP_PROPERTY_LIST));
+  /* Add DDS Security configuration to the QoS as a Property policy,
+   * used by the security plugins to get their proper settings. */
+  if (!(xqos->present & QP_PROPERTY_LIST) || xqos->property.value.n == 0)
```

Comment:
On consideration, I think it is reasonable to enforce the settings from the configuration for any participant that does not set the security properties explicitly, because that makes it much harder to accidentally create a non-secured participant with a configuration that specifies security.

I do wonder why the presence of any security property is enough to override them all. What is the (intended) effect of a partial security policy? I would expect it would be rejected as an inconsistent QoS on creation of the participant. @martinbremmer, I suppose you may know off the top of your head if that is indeed the case? If not, why not?

This is something worth noting in the documentation or, at least, comments.

Apart from that, it would make more sense to
* initialise the properties to (0,``NULL``) if they're not set
* then (always) check for presence of the security properties (even if it were just initialised) — this is a trivial check because the list is empty
* then add the properties if they are needed

that shrinks the code while making it simpler.

Reply:
> What is the (intended) effect of a partial security policy? I would expect it would be rejected as an inconsistent QoS on creation of the participant. @martinbremmer, I suppose you may know off the top of your head if that is indeed the case?

Using security properties from a user QoS is new. We only supported security settings through the configuration.
When creating a participant, the security properties are (will be) used to load the plugins and initialize them. This will fail if the properties are not correct or missing[1]. This in turn will fail the creation of the participant. At least, that's what I had in mind.

[1] Parts will be tested by ddsi (mostly related to plugins loading), parts will be tested by the plugins (mostly related to certificates and such).

Reply:
I have simplified `nn_xqos_mergein_security_config` and added code in `new_participant_guid` to check if all required security parameters are set when using qos properties for security configuration. With this check added, I think it is reasonable to override the security config if any security property is present, because the check prevents creating an unsecured participant accidentally by using the reserved `dds.sec.` prefix in a property name.